test/xbgpu/test_beamform.py

```diff
@@ -0,0 +1,114 @@
+            for beam in range(n_beams):
+                p = beam_pols[beam]
+                w = weights[:, beam]
+                w = w * np.exp(delays[:, beam] * channel * np.complex64(1j) * np.float32(np.pi))
+                for time in range(n_times):
+                    # .copy() is added because otherwise we get a warning that np.dot is
+                    # more efficient on C-contiguous arrays.
+                    accum = np.dot(in_c[:, time, p].copy(), w)
+                    out[frame, beam, channel, time, 0] = quant(accum.real)
+                    out[frame, beam, channel, time, 1] = quant(accum.imag)
+    return out
+
+
+@pytest.mark.combinations(
+    "n_frames, n_channels, n_times, n_antennas",
+    [1, 5],
+    [1, 128, 200, 1025],
+    [1, 128, 256, 321],
+    [1, 4, 19, 80],
+)
+def test_beamform(
+    context: AbstractContext,
+    command_queue: AbstractCommandQueue,
+    n_frames: int,
+    n_channels: int,
+    n_times: int,
+    n_antennas: int,
+) -> None:
+    """Test :class:`.Beamform` by comparing it to a CPU implementation."""
+    beam_pols = [0, 1, 0, 1, 1]
+    n_beams = len(beam_pols)
+
+    template = BeamformTemplate(context, beam_pols)
+    fn = template.instantiate(command_queue, n_frames, n_antennas, n_channels, n_times)
+    fn.ensure_all_bound()
+    h_in = fn.buffer("in").empty_like()
+    h_out = fn.buffer("out").empty_like()
+    h_weights = fn.buffer("weights").empty_like()
+    h_delays = fn.buffer("delays").empty_like()
+    assert h_in.shape == (n_frames, n_antennas, n_channels, n_times, N_POLS, COMPLEX)
+    assert h_out.shape == (n_frames, n_beams, n_channels, n_times, COMPLEX)
+    assert h_weights.shape == (n_antennas, n_beams)
+    assert h_delays.shape == (n_antennas, n_beams)
+
+    rng = np.random.default_rng(seed=123)
+    h_in[:] = rng.integers(-127, 127, h_in.shape)
+    scale = 2.0 / np.sqrt(n_antennas)
+    h_weights[:] = rng.uniform(-scale, scale, h_weights.shape) + 1j * rng.uniform(-scale, scale, h_weights.shape)
+    h_delays[:] = rng.uniform(-0.1, 0.1, h_delays.shape)
+    h_out.fill(0)
+    expected = beamform_host(h_in, h_weights, h_delays, np.array(beam_pols))
+
+    fn.buffer("out").zero(command_queue)
+    fn.buffer("in").set(command_queue, h_in)
+    fn.buffer("weights").set(command_queue, h_weights)
+    fn.buffer("delays").set(command_queue, h_delays)
+    fn()
+    fn.buffer("out").get(command_queue, h_out)
+
+    np.testing.assert_allclose(h_out, expected, atol=1)
+    # Ensure that the scale factor on the weights causes some clamping, but
+    # not too much. But only do it for larger test cases; in small tests cases
+    # it is hard to guarantee this.
+    if n_antennas > 1 and n_frames * n_channels * n_times > 1000:
+        clamped = np.sum(np.abs(h_out) == 127, axis=(0, 1, 2, 3)) / h_out[..., 0].size
+        assert 0.1 < clamped[0] < 0.9  # real
+        assert 0.1 < clamped[1] < 0.9  # imag
```

Comment:
> [...] in small *test cases it is hard to guarantee this.

Why is that?

Reply:
The data is random, and with small sample sizes the result is very sensitive to the seed. In the extreme case of only having 1 sample, it'll either be saturated or not, so `clamped[x]` will be either 0.0 or 1.0 and the test will definitely fail.